src/main/kotlin/org/rust/lang/core/resolve/ImplLookup.kt

```diff
@@ -91,24 +93,28 @@ class ImplLookup(
             return _ctx!!
         }
 
-    fun findImplsAndTraits(ty: Ty): Collection<RsTraitOrImpl> {
+    fun findImplsAndTraits(ty: Ty): Set<RsTraitOrImpl> {
         return findImplsAndTraitsCache.getOrPut(project, freshen(ty)) { rawFindImplsAndTraits(ty) }
     }
 
-    private fun rawFindImplsAndTraits(ty: Ty): Collection<RsTraitOrImpl> {
-        return when (ty) {
-            is TyTypeParameter -> ty.getTraitBoundsTransitively().map { it.element }
-            is TyTraitObject -> BoundElement(ty.trait).flattenHierarchy.map { it.element }
+    private fun rawFindImplsAndTraits(ty: Ty): Set<RsTraitOrImpl> {
+        val implsAndTraits = mutableSetOf<RsTraitOrImpl>()
+        when (ty) {
+            is TyTypeParameter -> ty.getTraitBoundsTransitively().mapTo(implsAndTraits) { it.element }
+            is TyTraitObject -> BoundElement(ty.trait).flattenHierarchy.mapTo(implsAndTraits) { it.element }
             is TyFunction -> {
-                findSimpleImpls(ty) + fnTraits
+                implsAndTraits += findSimpleImpls(ty)
+                implsAndTraits += fnTraits
+            }
+            is TyUnknown -> Unit
+            else -> {
+                implsAndTraits += findDerivedTraits(ty)
+                implsAndTraits += findSimpleImpls(ty)
+                getHardcodedImpls(ty).mapTo(implsAndTraits) { it.element }
             }
-            is TyUnknown -> emptyList()
-            else -> concatLists(
-                findDerivedTraits(ty),
-                getHardcodedImpls(ty).map { it.element },
-                findSimpleImpls(ty)
-            )
         }
+        implsAndTraits += findImplicitImpls(ty)
+        return implsAndTraits
```

Comment:
👍

Reply:
In general, I think we should switch our style from recklessly concatenating read only collections to  the pattern of creating a mutable collection, filling it with values and then freezing it. There are helper functions like `buildList` which could help here, though in this particular case just creating a mutable set seems clearer.